### ISSUE TITLE
Parser violates Postel's Law

### ISSUE THREAD

**Issue by gkuenning** [dateutil/dateutil#480]
The dateutil package is too eager to reject nonconforming VCALENDAR-format files.  The particular issue I encountered is related to time zones.  RFC 5545 clearly states that the DTSTART field is required.  However, there are situations where DTSTART isn't strictly necessary because the zone in question doesn't observe daylight-savings time and never has.

For example, I have a VCALENDAR file that was generated by a program that omits DTSTART for such zones.  Here's an example:

BEGIN:VTIMEZONE
TZID:America/Phoenix
X-LIC-LOCATION:America/Phoenix
BEGIN:DAYLIGHT
TZOFFSETFROM:-0700
TZOFFSETTO:-0700
TZNAME:Ariz
END:DAYLIGHT
BEGIN:STANDARD
TZOFFSETFROM:-0700
TZOFFSETTO:-0700
TZNAME:Ariz
END:STANDARD
END:VTIMEZONE

Clearly, this file violates RFC 5445, and I have reported that fact to the program's maintainer (who will fix the problem soon).  Nevertheless, feeding an ICS file to datetime._parse_rfc with this error causes a ValueError exception, which makes the VCALENDAR file unreadable.

In keeping with Postel's law ("Be conservative in what you do, be liberal in what you accept from others"), _parse_rfc should attempt to accept VCALENDAR files whenever it is possible to make sense of them.  Thus, for example:

    if not founddtstart:
        rrulelines.append('DTSTART:19000101T020000')

The above could be improved a bit, for example by still rejecting entries in which the standard and daylight sections had different offsets (although even then it seems valid to assume a DTSTART in the distant past).

Although the dtstart issue is the one that prompted this report, I also noticed the following in _parse_rfc:

    if name == "BEGIN":
        if value in ("STANDARD", "DAYLIGHT"):
            # Process component
            pass
        else:
            raise ValueError("unknown component: "+value)

Again, there's an opportunity to be more robust here.  One could issue a warning message, but then ignore the unknown component.

In both cases (and I suspect numerous others), halting parsing is an extreme response to various errors, since it leaves the user of the package with no way to process a nonconforming file.  That's especially problematic since VCALENDAR files are generated by so many different programs, many of which are written by programmers who haven't bothered to read RFC 5445--or who have read it but then made some minor mistake that produces broken files.

**Comment by pganssle** [comment]
I will admit that I am not a fan of this principle as a general rule, as it tends to propagate deviation from a standard. Errors are a signal that something is wrong, and the only reason your non-conforming generation app knows that their code was not following the standard is because a parser that *is* following the standard (dateutil) threw an error. I think it is not good for the standards ecosystem in general if the rule is that parsers are liberal about enforcing standards. Consider [this memo](https://tools.ietf.org/html/draft-thomson-postel-was-wrong-00) about the harmful consequences of Postel's maxim. I certainly do not think that dateutil should attempt to *silently* fix non-compliant strings.

That said, where possible I like to expose seams so that people are not left with the choice of something working perfectly or not at all, so I will look at this and consider ways that someone could extend one of these parsers to allow for non-conforming strings, or possibly to throw meaningful error types that might allow programmers to recover from the error.

However, I have several other problems with your specific examples. First, I don't understand what exactly your input is here. `dateutil` does not offer facilities for parsing `VCALENDAR` strings. There are two RFC 2445 (**not** currently 5455, mind you)-style parsers, `tzical` and `rrulestr`.  `_parse_rfc` is not part of the public interface - if you are calling it directly, you do so at your own risk.

With regards to the specific issue of `DTSTART`, I'll have to look closer to see if it's even meaningful to omit that parameter. In the case of time zones it might happen to work *if* there's only one rule in the zone, and even then I don't know how that interacts with the `DTSTART` parameter of `RRULE`. (As I mentioned above, the default behavior will remain RFC compliant, but it may be possible to eventually parse to an intermediate representation first which is then validated, and downstream users can subclass to skip certain parts of the validation).

**Comment by gkuenning** [comment]
I agree with you about the problems with Postel's Law, and also appreciate your sensible comments on all-or-nothing interfaces.  Perhaps there could be a strict-vs-loose option (defaulting to strict)?

As to how the problem arose, I have a program that uses the vobject interface to read a vcalendar file (vobject.readComponents); that package apparently invokes dateutil and eventually _parse_rfc.  The exception gets thrown there and is caught and re-thrown by vobject.transformToNative.

**Comment by pganssle** [comment]
I don't think there will be a "strict vs. loose" option. More likely what I said above, parse to an intermediate representation if that's feasible, then pass through a validation step, then return. Downstream users can manipulate the intermediate representation before it hits the validation step if they want to allow specific non-compliant behaviors.
